kore/src/Kore/Step/Simplification/Overloading.hs

```diff
@@ -27,113 +27,154 @@ import Kore.Internal.Pattern
+            termMerger firstTerm' secondTerm'
+        Left message -> Monad.Trans.lift $
+            explainAndReturnBottom message firstTerm secondTerm
+        Right Nothing -> empty
+
+{- |
+ Tests whether the pair of terms can be coerced to have the same constructors
+ at the top, and, if so, returns the thus obtained new pair.
+
+ See <https://github.com/kframework/kore/blob/master/docs/2019-08-27-Unification-modulo-overloaded-constructors.md>
+
+ If the overloading rules are not appliable, it would return @Nothing@.
+ However, if it detects an unification failure, it would return a reason
+ for the failure which can be used as an explanation for it.
+
+ Note: this algorithm is used for both matching and unification; that is why
+ the first and second terms in a pair are not interchangeable.
+-}
+unifyOverloading
+    :: MonadSimplify unifier
+    => SimplifierVariable variable
+    => Pair (TermLike variable)
+    -> unifier
+        (Either (Doc ()) (Maybe (Pair (TermLike variable))))
+unifyOverloading termPair = case termPair of
+    Pair
+        (Inj_ inj@Inj { injChild = App_ firstHead firstChildren })
+        secondTerm@(App_ secondHead _)
+        -> flipPairBack <$> unifyOverloadingVsOverloaded
+            secondHead
+            secondTerm
+            (Application firstHead firstChildren)
+            inj { injChild = () }
+    Pair
+        firstTerm@(App_ firstHead _)
+        (Inj_ inj@Inj { injChild = App_ secondHead secondChildren })
+        -> unifyOverloadingVsOverloaded
+            firstHead
+            firstTerm
+            (Application secondHead secondChildren)
+            inj { injChild = () }
+    Pair
+        (Inj_ inj@Inj { injChild = App_ firstHead firstChildren })
+        (Inj_ Inj { injChild = App_ secondHead secondChildren })
+        -> unifyOverloadingCommonOverload
+            (Application firstHead firstChildren)
+            (Application secondHead secondChildren)
+            inj { injChild = () }
+    (Pair (App_ firstHead _) (Inj_ Inj { injChild })) ->
+        notUnifiableTest firstHead injChild
+    (Pair (Inj_ Inj { injChild }) (App_ secondHead _)) ->
+        notUnifiableTest secondHead injChild
+    _ -> return $ Right Nothing
+  where
+    flipPairBack (Right (Just (Pair x y))) = Right (Just (Pair y x))
+    flipPairBack p = p
+    notUnifiableTest termHead child = do
+        OverloadSimplifier { isOverloaded } <- Simplifier.askOverloadSimplifier
+        if isOverloaded termHead
+        then case notUnifiableType child of
+            Just typeName -> return $ Left typeName
+            Nothing -> return $ Right Nothing
+        else return $ Right Nothing
+
+unifyOverloadingCommonOverload
+    :: MonadSimplify unifier
+    => SimplifierVariable variable
+    => Application Symbol (TermLike variable)
+    -> Application Symbol (TermLike variable)
+    -> Inj ()
+    -> unifier
+        (Either (Doc ()) (Maybe (Pair (TermLike variable))))
```

Comment:
Could you make a custom UnificationResult type instead of Either+Maybe?

Reply:
I agree, this should at least be `unifier (Either UnifyOverloadingError (Pair (TermLike variable)))`. Another option, which should minimize the need for `case` and `if _ then _ else _`, is `ExceptT UnifyOverloadingError unifier (Pair (TermLike variable))`. You could define a type synonym such as,

```.hs
type UnifyOverloadingResult a = Either UnifyOverloadingError (Pair a)
-- or:
type UnifyOverloading unifier a = ExceptT UnifyOverloadingError unifier (Pair a)
```

or a `newtype`. If you want to go the `newtype` route, please also define functions to make that type as easy to deconstruct as `either` makes `Either` easy to deconstruct.

`UnifyOverloadingError` should be a sum type so we can match on the result, at least for testing.